Objects/typevarobject.c

```diff
@@ -0,0 +1,1724 @@
+// TypeVar, TypeVarTuple, and ParamSpec
+#include "Python.h"
+#include "pycore_object.h"  // _PyObject_GC_TRACK/UNTRACK
+#include "pycore_typevarobject.h"
+#include "pycore_unionobject.h"   // _Py_union_type_or
+#include "structmember.h"
+
+/*[clinic input]
+class typevar "typevarobject *" "&_PyTypeVar_Type"
+class paramspec "paramspecobject *" "&_PyParamSpec_Type"
+class paramspecargs "paramspecattrobject *" "&_PyParamSpecArgs_Type"
+class paramspeckwargs "paramspecattrobject *" "&_PyParamSpecKwargs_Type"
+class typevartuple "typevartupleobject *" "&_PyTypeVarTuple_Type"
+class typealias "typealiasobject *" "&_PyTypeAlias_Type"
+class Generic "PyObject *" "&PyGeneric_Type"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=aa86741931a0f55c]*/
+
+typedef struct {
+    PyObject_HEAD
+    const char *name;
+    PyObject *bound;
+    PyObject *evaluate_bound;
+    PyObject *constraints;
+    PyObject *evaluate_constraints;
+    bool covariant;
+    bool contravariant;
+    bool infer_variance;
+} typevarobject;
+
+typedef struct {
+    PyObject_HEAD
+    const char *name;
+} typevartupleobject;
+
+typedef struct {
+    PyObject_HEAD
+    const char *name;
+    PyObject *bound;
+    bool covariant;
+    bool contravariant;
+    bool infer_variance;
+} paramspecobject;
+
+typedef struct {
+    PyObject_HEAD
+    const char *name;
+    PyObject *type_params;
+    PyObject *compute_value;
+    PyObject *value;
+} typealiasobject;
+
+#include "clinic/typevarobject.c.h"
+
+static PyObject *
+call_typing_func_object(const char *name, PyObject *args)
+{
+    PyObject *typing = PyImport_ImportModule("typing");
+    if (typing == NULL) {
+        return NULL;
+    }
+    PyObject *func = PyObject_GetAttrString(typing, name);
+    if (func == NULL) {
+        Py_DECREF(typing);
+        return NULL;
+    }
+    PyObject *result = PyObject_CallObject(func, args);
+    Py_DECREF(func);
+    Py_DECREF(typing);
+    return result;
+}
+
+static PyObject *
+type_check(PyObject *arg)
+{
+    // Calling typing.py here leads to bootstrapping problems
+    if (Py_IsNone(arg)) {
+        return Py_NewRef(Py_TYPE(arg));
+    }
+    // TODO: real error message
+    PyObject *args = PyTuple_Pack(2, arg, Py_None);
+    if (args == NULL) {
+        return NULL;
+    }
+    PyObject *result = call_typing_func_object("_type_check", args);
+    Py_DECREF(args);
+    return result;
+}
+
+/*
+ * Return a typing.Union. This is used as the nb_or (|) operator for
+ * TypeVar and ParamSpec. We use this rather than _Py_union_type_or
+ * (which would produce a types.Union) because historically TypeVar
+ * supported unions with forward references, and we want to preserve
```

Comment:
clarify that we're talking about simple strings here, not lazy evaluation or something

```suggestion
 * supported unions with string forward references, and we want to preserve
```